### ISSUE TITLE
BPF tail calls exceed more than the 32 function limit

### ISSUE THREAD

**Issue by shpalani** [microsoft/ebpf-for-windows#2496]
### Describe the bug

Logic:
bpf_tail_calls should not exceed more than 32 (MAX_TAIL_CALL_CNT)

This function ebpf_program_set_tail_call() checks the number of tail calls, and fails the tail call if the number exceeds MAX_TAIL_CALL_CNT

In #2482, I tested with 35 bpf_tail_call function calls for bind hook. The test case passed, instead of failing.

CI/CD pipeline: https://github.com/microsoft/ebpf-for-windows/actions/runs/5007785677/jobs/8974850728?pr=2482

### OS information

Windows 10 and above

### Steps taken to reproduce bug

1. Patch the diff 
[tail_call_diff.txt](https://github.com/microsoft/ebpf-for-windows/files/11523795/tail_call_diff.txt)
2. Run api_test.exe

### Expected behavior

bindmonitor_mt_tailcall_native_test test case in api_test.exe should fail.

### Actual outcome

bindmonitor_mt_tailcall_native_test test case in api_test.exe passed.

### Additional details

When I changed 
#define MAX_TAIL_CALL_PROGS 45
The test case with bpf_tail_call failed at 42. This shows that the max number of allowed bpf tail call is 42, instead of 32.

**Comment by shpalani** [comment]
**Tested on latest ebpf binaries with the below diff.  The problem is reproducible.**

- With **35 tail calls**,

D:\ebpf-for-windows>**git diff**
diff --git a/tests/sample/bindmonitor_mt_tailcall.c b/tests/sample/bindmonitor_mt_tailcall.c
index 41809528..5376949a 100644
--- a/tests/sample/bindmonitor_mt_tailcall.c
+++ b/tests/sample/bindmonitor_mt_tailcall.c
@@ -24,7 +24,7 @@ struct bpf_map_def bind_tail_call_map = {
     .type = BPF_MAP_TYPE_PROG_ARRAY,
     .key_size = sizeof(uint32_t),
     .value_size = sizeof(uint32_t),
-    .max_entries = MAX_TAIL_CALL_CNT};
+    .max_entries = MAX_TAIL_CALL_CNT + 3};

 SEC("bind")
 bind_action_t
@@ -78,15 +78,18 @@ DEFINE_BIND_TAIL_FUNC(27)
 DEFINE_BIND_TAIL_FUNC(28)
 DEFINE_BIND_TAIL_FUNC(29)
 DEFINE_BIND_TAIL_FUNC(30)
+DEFINE_BIND_TAIL_FUNC(31)
+DEFINE_BIND_TAIL_FUNC(32)
+DEFINE_BIND_TAIL_FUNC(33)

 // This line verifies that the BindMonitor_Caller prototype is correct by declaring a bind_hook_t
 // variable with the same name as the first tail call function.
 // This line is optional.
 bind_hook_t BindMonitor_Caller;

-SEC("bind/31")
+SEC("bind/34")
 bind_action_t
-BindMonitor_Callee31(bind_md_t* ctx)
+BindMonitor_Callee34(bind_md_t* ctx)
 {
     // This function is the last tail call function for the bind hook.
     // This function returns BIND_PERMIT to allow the bind request to proceed.
diff --git a/tests/stress/km/stress_tests_km.cpp b/tests/stress/km/stress_tests_km.cpp
index 66f88e84..1c9439db 100644
--- a/tests/stress/km/stress_tests_km.cpp
+++ b/tests/stress/km/stress_tests_km.cpp
@@ -1118,7 +1118,7 @@ _set_up_tailcall_program(bpf_object* object, const std::string& map_name)
     LOG_VERBOSE("({}) Opened fd:{} for map:{}", __func__, prog_map_fd, map_name.c_str());

     // Set up tail calls.
-    for (int index = 0; index < MAX_TAIL_CALL_CNT; index++) {
+    for (int index = 0; index < MAX_TAIL_CALL_CNT + 3; index++) {
         try {
             std::string bind_program_name{"BindMonitor_Callee"};
             bind_program_name += std::to_string(index);
(END)

- Run was successful:

PS C:\Temp> **.\ebpf_stress_tests_km.exe -tt=2 -td=2 -vo=true bindmonitor_tail_call_invoke_program_test**
...
Thread[1] - binding to port:8990
Thread[1] bind success to port:8990
Thread[1] Done.
Thread[0] - binding to port:8989
Thread[0] bind success to port:8989
Thread[0] - binding to port:8989
Thread[0] bind success to port:8989
Thread[0] Done.
===============================================================================
**All tests passed (**124492 assertions in 1 test case)

- Traces show the bind was successful, with 35 tail calls.
 
[tail_call_debug.zip](https://github.com/microsoft/ebpf-for-windows/files/12236642/tail_call_debug.zip)

**Comment by shpalani** [comment]
Status:
1. **User mode:** Added a test case  in **unit_test.exe**. The test case fails correctly, as the number of tail calls are more than 32.
2. **Kernel mode**: Added a test case in **api_test.exe.** The test case passes, even when the number of tail calls are more than 32.
3. **Kernel mode:** Extended the stress test case in **ebpf_stress_tests_km.exe**. The test case passes, even when the number of tail calls are more than 32.

In short, the problem is observed in kernel mode test cases.

**Comment by shpalani** [comment]
**Root Cause Analysis:**

- The Program handle is incorrect while attaching in **ebpf_program_attach_by_fd().** This is due to 
      - This function **_get_handle_from_file_descriptor(fd_t fd)** returns an incorrect program handle for a given fd.
- Only 25 tail call programs are invoked, starting from BindMonitor_Callee9 - BindMonitor_Callee34. Since the number 25 < MAX_TAIL_CALL_CNT, and the last tail BindMonitor_Callee34 program containing BIND_PERMIT is invoked, the test passes.

Conclusion: The bug is in the incorrect program attachment, and not in the max tail call handling.

**Logs:**
PS C:\Temp>.\api_test.exe bindmonitor_mt_tailcall_native_test

Filters: "bindmonitor_mt_tailcall_native_test"
Randomness seeded to: 2761246397
Loaded native module bindmonitor_mt_tailcall.sys, [fd=3], [count of programs=[36]
Loaded program bindmonitor_mt_tailcall.sys, [fd=4]

program name [**BindMonitor_Caller**], program_fd: [39]
ebpf_program_attach_by_fd: program handle [**748**], program_fd=[**39**]

Successfully loaded program
bpf_object__find_program_by_name succeeded for [BindMonitor_Caller], fd=[39]

**From traces:**
1. Program handle 748 belongs to BindMonitor_Callee9 function.
2. Program handle 588 belongs to BindMonitor_Caller function.
3. _get_handle_from_file_descriptor() returns 748 for fd 39.
4. Hence program attachment is done to program handle which belongs to BindMonitor_Callee9 tail function.


20:58:52.080 [EbpfForWindowsProvider]ebpf_handle_create
[6]10A8.203C::?2023?-?08?-?11 20:58:52.080 [EbpfForWindowsProvider]ebpf_handle_create: returning handle,748
[6]10A8.203C::?2023?-?08?-?11 20:58:52.080 [EbpfForWindowsProvider]ebpf_handle_create returned success
[6]10A8.203C::?2023?-?08?-?11 20:58:52.080 [EbpfForWindowsProvider]ebpf_program_create_and_initialize: Program name,**BindMonitor_Callee9**
[6]10A8.203C::?2023?-?08?-?11 20:58:52.080 [EbpfForWindowsProvider]**ebpf_program_create_and_initialize: handle , program handle,748**
[6]10A8.203C::?2023?-?08?-?11 20:58:52.080 [EbpfForWindowsProvider]ebpf_program_create_and_initialize returned success

...
[6]10A8.203C::?2023?-?08?-?11 20:58:52.048 [EbpfForWindowsProvider]ebpf_handle_create
[6]10A8.203C::?2023?-?08?-?11 20:58:52.048 [EbpfForWindowsProvider]ebpf_handle_create: returning handle,**588**
[6]10A8.203C::?2023?-?08?-?11 20:58:52.048 [EbpfForWindowsProvider]ebpf_handle_create returned success
[6]10A8.203C::?2023?-?08?-?11 20:58:52.048 [EbpfForWindowsProvider]ebpf_program_create_and_initialize: Program name,**BindMonitor_Caller**
[6]10A8.203C::?2023?-?08?-?11 20:58:52.048 [EbpfForWindowsProvider]ebpf_program_create_and_initialize: handle , program handle,588
[6]10A8.203C::?2023?-?08?-?11 20:58:52.048 [EbpfForWindowsProvider]ebpf_program_create_and_initialize returned success
..
[6]10A8.203C::?2023?-?08?-?11 20:58:52.084 [EbpfForWindowsProvider]ebpf_link_create
[6]10A8.203C::?2023?-?08?-?11 20:58:52.084 [EbpfForWindowsProvider]eBPF object initialized,0xFFFFC20FEF79CEE0,2
[6]10A8.203C::?2023?-?08?-?11 20:58:52.084 [EbpfForWindowsProvider]ebpf_link_create returned success
[6]10A8.203C::?2023?-?08?-?11 20:58:52.084 [EbpfForWindowsProvider]ebpf_link_attach_program
[6]10A8.203C::?2023?-?08?-?11 20:58:52.084 [EbpfForWindowsProvider]ebpf_program_attach_link
[6]10A8.203C::?2023?-?08?-?11 20:58:52.084 [EbpfForWindowsProvider]**ebpf_program_attach_link: program name,BindMonitor_Callee9**
[6]10A8.203C::?2023?-?08?-?11 20:58:52.084 [EbpfForWindowsProvider]ebpf_program_attach_link: program link count,1

**Comment by shpalani** [comment]
Further digging,

The order in ebpfapi program vector is 
**BindMonitor_Callee0 - 34 and the last one is BindMonitor_Caller.**
Vs
The order in ebpfcore is
**BindMonitor_Caller , BindMonitor_Callee0 - 34**
 
During initialization, programs[i] and program_handles[i] are mismatched ie, offset by index 1.
_initialize_ebpf_programs_native(
{
         ...
        ebpf_program_t* program = programs[i];
        program->fd = _create_file_descriptor_for_handle(program_handles[i]);
        ...
}

Hence there is incorrect attachment.

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/shpalan/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

<link rel=themeData
href="file:///C:/Users/shpalan/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/shpalan/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">

<style>

</style>

</head>

<body lang=EN-US style='tab-interval:.5in;word-wrap:break-word'>




Ebpfapi | EbpfCore: ebpf_program_create_and_initialize()
-- | --
612  [BindMonitor_Callee0] | BindMonitor_Caller   612
616 [BindMonitor_Callee1] | BindMonitor_Callee0  616
624 [BindMonitor_Callee10] | BindMonitor_Callee1   624
628[BindMonitor_Callee11] | 628 BindMonitor_Callee10
BindMonitor_Callee12   – 19       BindMonitor_Callee19:   660 | 632 BindMonitor_Callee11
664 BindMonitor_Callee2 | BindMonitor_Callee12   – 19       BindMonitor_Callee19:   664
668 BindMonitor_Callee20 | 668 BindMonitor_Callee2
672 BindMonitor_Callee21 | 672 BindMonitor_Callee20
BindMonitor_Callee22-29       BindMonitor_Callee29:   704 | 676 BindMonitor_Callee21
708 BindMonitor_Callee3 | BindMonitor_Callee22-29           BindMonitor_Callee29:   708
712 BindMonitor_Callee30 | 712 BindMonitor_Callee3
716 BindMonitor_Callee31 | 716 BindMonitor_Callee30
BindMonitor_Callee32   – 34       BindMonitor_Callee32:   720   BindMonitor_Callee34:   728 | 720 BindMonitor_Callee31
732 BindMonitor_Callee4 | BindMonitor_Callee   32- 34       BindMonitor_Callee34:   732
BindMonitor_Callee5   – 8       BindMonitor_Callee5:   736   BindMonitor_Callee8:   748 | 736 BindMonitor_Callee4
752 BindMonitor_Callee9 | BindMonitor_Callee5   – 8       BindMonitor_Callee8:   752
756 [BindMonitor_Caller] | 756   BindMonitor_Callee9




</body>

</html>

**Comment by shpalani** [comment]
With the fix of https://github.com/microsoft/ebpf-for-windows/pull/2784, it was resolved.

But a new issue is found:
**Problem:** Only 31 tail calls are allowed.

MAX_TAIL_CALL_CNT = 32;

**Root cause:** The below EBPF_NO_MORE_TAIL_CALLS returns at 31 (MAX_TAIL_CALL_CNT - 1)

diff --git a/libs/execution_context/ebpf_program.c b/libs/execution_context/ebpf_program.c
index af03d985..b3245071 100644
--- a/libs/execution_context/ebpf_program.c
+++ b/libs/execution_context/ebpf_program.c
@@ -1356,6 +1356,11 @@ ebpf_program_set_tail_call(_In_ const ebpf_program_t* next_program)

     if (**state->count == (MAX_TAIL_CALL_CNT - 1**)) {
         EBPF_OBJECT_RELEASE_REFERENCE(&((ebpf_program_t*)next_program)->object);
+        EBPF_LOG_MESSAGE_UINT64(
+            EBPF_TRACELOG_LEVEL_ERROR,
+            EBPF_TRACELOG_KEYWORD_PROGRAM,
+            "State count exceeded maximum for program tail call, Returning EBPF_NO_MORE_TAIL_CALLS",
+            state->count);
         return EBPF_NO_MORE_TAIL_CALLS;
     }

 

(END)
--
From the trace,

[29]1F48.1A4C::?2023?-?08?-?24 19:47:31.777 [EbpfForWindowsProvider]Calling Tail call index [x = 30]
[29]1F48.1A4C::?2023?-?08?-?24 19:47:31.777 [EbpfForWindowsProvider]S**tate count exceeded maximum for program tail call, Returning EBPF_NO_MORE_TAIL_CALLS,31**
[29]1F48.1A4C::?2023?-?08?-?24 19:47:31.777 [EbpfForWindowsProvider]Tail call failed at index 31

**Comment by shpalani** [comment]
Closing this task (with the analysis, root cause identified, and work in progress for the solution),  as it has been fixed recently by https://github.com/microsoft/ebpf-for-windows/pull/2784.
